### PR TITLE
apiextensions: [alpha] implement defaulting

### PR DESCRIPTION
**Pull request by sttts**
Based on https://github.com/kubernetes/kubernetes/pull/64558

- add `CustomResourceDefaulting` feature gate.
- implement defaulting via go-openapi validation.

```release-note
Added CRD defaulting via OpenAPI default fields in the validation schema, enabled via CustomResourceDefaulting feature gate.
```

### CODE REVIEW

**Comment by lavalamp** [comment]
I think we need to give validation errors instead of pruning fields. Since it's a new feature I think we can still do this for CRDs.

**Comment by sttts** [comment]
> I think we need to give validation errors instead of pruning fields. Since it's a new feature I think we can still do this for CRDs.

Technically this is easy to do. But do we really want that unconditionally? The behaviour will be different than for native resources. We had decided to replicate pruning for CRDs in the Sig meeting a couple of months ago to follow our existing conventions.

Also, if we rejected unknown fields, do we also want that for the storage decoder? Or do we want to prune in that case?

I totally agree that kubectl as an interactive tool should reject unknown fields. That behaviour could be implemented with some CreateOption, i.e. a strict mode, without changing the existing API convention of pruning.

**Comment by lavalamp** [comment]
> But do we really want that unconditionally? 

Yes

> The behaviour will be different than for native resources. 

Great!

> We had decided to replicate pruning for CRDs in the Sig meeting a couple of months ago to follow our existing conventions.

Was I there and conscious?

> Also, if we rejected unknown fields, do we also want that for the storage decoder? Or do we want to prune in that case?

That's a good point, the storage decoder should probably prune.

> I totally agree that kubectl as an interactive tool should reject unknown fields. That behaviour could be implemented with some CreateOption, i.e. a strict mode, without changing the existing API convention of pruning.

Strongly disagree that the client should be responsible for this behavior. You should get the same validation behavior whether you use python, curl or kubectl.

**Comment by lavalamp** [comment]
OK, I forgot about a case (new manifest getting sent to old server) that @liggitt reminded me of.

I think the correct solution is to add micro-versions, so apiserver can tell the difference between "wrong" and "for a version I don't know about".

However in the interest of expediency I guess I have to admit that we probably should prune by default and give validation errors on a --strict option or some such. :'(

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go [hunk 1]
```diff
@@ -51,6 +51,9 @@ type CustomResourceDefinitionSpec struct {
 	// by GA > beta > alpha, and then by comparing major version, then minor version. An example sorted list of
 	// versions: v10, v2, v1, v11beta2, v10beta3, v3beta1, v12alpha1, v11alpha2, foo1, foo10.
 	Versions []CustomResourceDefinitionVersion `json:"versions,omitempty" protobuf:"bytes,7,rep,name=versions"`
+	// Prune enables pruning of unspecified fields. This required validation to be enabled.
+	// Note: this will default to true in version v1.
```
Comment (deads2k) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go]: In the next version.   Also, I think we should require this to enable versioning beyond the null conversion.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 2]
```diff
@@ -772,8 +773,8 @@ type unstructuredSchemaValidator struct {
 	dropInvalidMetadata bool
 	validator           *validate.SchemaValidator
 	// empty or a top-level field name like "status" the validator should be applied to
-	validatedField string
-	defaulting     bool
+	validatedField      string
+	defaulting, pruning bool
 }
 
 func (v *unstructuredSchemaValidator) apply(u *unstructured.Unstructured) error {
```
Comment (deads2k) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: For those of us following along at home.  When does this get called?  In a flow that goes 
 1. decode to internal (or whatever that is for CR)
 2. call mutating admission
     1. convert to external for webhook
     2. convert back to internal
     3. repeat for all webhooks
 3. call rest strategy
 4. call prepareforOperation
 5. call validating admission
 6. call validate
 7. convert to storage version
 8. serialize to storage

I'd expect to have values pruned before mutating admission, after webhook patches are applied, and before validating admission, just .

If you don't do it before mutating admission, the first mutating webhook may get extraneous values.  If you don't do it after mutating admission and before validating admission, validation admission webhooks may get bad values.  If you don't do it before storage, a bad converter might cause you store bad values.

**Comment by spiffxp** [comment]
ping, needs rebase

**Comment by spiffxp** [comment]
/milestone v1.13
I'm adding this to the v1.13 milestone since it relates to https://github.com/kubernetes/features/issues/575 which is currently planned for v1.13

**Comment by shubheksha** [comment]
@nikopen, can we remove the milestone from this PR as kubernetes/enhancements#575 isn't tracked anymore.

**Pull request closed without merging**